src/ringmesh/io/geomodel_builder_resqml.cpp

```diff
@@ -0,0 +1,968 @@
+                    new double[pointCount * 3] );
+                tri_set->getXyzPointsOfPatch( patch, &xyzPoints[0] );
+
+                std::vector< vec3 > points( pointCount, vec3() );
+                for( auto i : range( pointCount ) )
+                {
+                    points[i] = vec3( xyzPoints[i * 3], xyzPoints[i * 3 + 1],
+                        xyzPoints[i * 3 + 2] );
+                }
+
+                const unsigned int triangleCount =
+                    tri_set->getTriangleCountOfPatch( patch );
+
+                std::vector< index_t > trgls( triangleCount * 3, 0 );
+
+                tri_set->getTriangleNodeIndicesOfPatch( patch, &trgls[0] );
+                for( auto& node : trgls )
+                {
+                    node -= (index_t) global_point_count;
+                }
+
+                std::vector< index_t > trgls_ptr( triangleCount + 1, 0 );
+                for( auto i : range( trgls_ptr.size() ) )
+                {
+                    trgls_ptr[i] = i * 3;
+                }
+
+                const gmme_id children = builder_.topology.create_mesh_entity(
+                    Surface3D::type_name_static() );
+
+                builder_.geology.add_parent_children_relation(
+                    interface_id, children );
+
+                builder_.geometry.set_surface_geometry(
+                    children.index(), points, trgls, trgls_ptr );
+
+                for( auto prop_index :
+                    range( tri_set->getValuesPropertySet().size() ) )
+                {
+                    AbstractValuesProperty* propVal =
+                        tri_set->getValuesPropertySet()[prop_index];
+                    const Surface3D& cur_surf =
+                        geomodel_.surface( children.index() );
+
+                    const gsoap_resqml2_0_1::resqml2__IndexableElements
+                        attachment = propVal->getAttachmentKind();
+                    if( attachment
+                        == gsoap_resqml2_0_1::
+                               resqml2__IndexableElements__nodes )
+                    {
+                        read_property( *propVal, patch,
+                            cur_surf.vertex_attribute_manager() );
+                    }
+                    else if( attachment
+                             == gsoap_resqml2_0_1::
+                                    resqml2__IndexableElements__cells )
+                    {
+                        read_property( *propVal, patch,
+                            cur_surf.polygon_attribute_manager() );
+                    }
+                }
+
+                global_point_count += pointCount;
+            }
+        }
+
+        builder_.build_lines_and_corners_from_surfaces();
+        builder_.build_regions_from_lines_and_surfaces();
+        builder_.geology.build_contacts();
+        return true;
+    }
+
+    namespace
+    {
+        bool read_volume_rep( VolumeMesh3D& mesh,
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this function to reduce its Cognitive Complexity from 33 to the 25 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3AS3776)